Fix the workspace calculation in miopenConvolutionABBackwardWeightsGetWorkSpaceSize

JonathanLichtnerAMD opened the pull request:
This PR fixes the workspace calculation in miopenConvolutionABBackwardWeightsGetWorkSpaceSize.  I would have preferred to use GetWorkspaceSize from CK to do this, but that would have required a handle and other things, which became very messy.  Instead, I used the formula that @bartekxk indicated: `G * K * C * D * H * W * sizeof(data type)`.

Note that this change _mostly_ resolves the crash we were seeing with the following MIOpenDriver command:

```MIOPEN_ENABLE_LOGGING_CMD=1 MIOPEN_ENABLE_LOGGING=1 MIOPEN_LOG_LEVEL=6 MIOPEN_FIND_ENFORCE=3 MIOPEN_FIND_MODE=3 ./bin/MIOpenDriver convbfp16 -n 256 -c 3 --in_d 2 -H 14 -W 14 -k 1280 --fil_d 2 -y 14 -x 14 --pad_d 0 -p 0 -q 0 --conv_stride_d 2 -u 14 -v 14 --dilation_d 1 -l 1 -j 1 --spatial_dim 3 -m conv -g 1 -F 4 -t 1```

However, I still observe a crash if we run this command with MIOPEN_DEBUG_CHECK_SUB_BUFFER_OOB_MEMORY_ACCESS equal to 1 or 2, so that still needs some investigation.

General TODOs:
- [x] investigate crash with MIOPEN_DEBUG_CHECK_SUB_BUFFER_OOB_MEMORY_ACCESS
- [x] add a unit test to reproduce the issue
- [x] some unit tests are failing because the mallocs are too large (~96GB) on MI200
- [x] check where GetNumBytes() is used to see if not accounting for group size is safe

Testing TODOs:
- [x] try fix on Fremont reproduction
- [x] run on a large subset of backward weights shapes (with large Ks)
- [ ] maybe run on Fremont shapes (not needed)
- [ ] run on MI100 CI (not done since MI100 CI was broken)
- [ ] run on MI300 CI (not done since MI300 CI was broken)

JonathanLichtnerAMD committed:
Log the CK kernel name when it is executed

JonathanLichtnerAMD committed:
Swap C and K since the the input and output tensors have been swapped

The logging was showing that C was K and vice versa.

JonathanLichtnerAMD committed:
Add G to the MIOPEN log for miopenConvolutionABBackwardWeightsGetWorkSpaceSize

JonathanLichtnerAMD committed:
Fix spatial dim usage in miopenConvolutionABBackwardWeightsGetWorkSpaceSize

The code was using the tensor dimensionality but that is not the same
as the spatial dimensionality.  This worked for getting K and C, but it
would not work for getting D, H, and W via GetNCDHW.

JonathanLichtnerAMD committed:
Fix the workspace calculation in miopenConvolutionABBackwardWeightsGetWorkSpaceSize

We log N for the fun of it.

JonathanLichtnerAMD commented:
Note that some unit tests are failing because the malloc for the work space size is too large:

        10793 - Full/GPU_GroupConv3D_BackwardWeights_FP16.GroupConv3D_BackwardWeights_half_Test/( G:1 N:6 C:448 K:896 D:3 H:118 W:182 z:1 y:1 x:1 pad.z:0 pad.y:0 pad.x:0 stride.z:1 stride.y:2 stride.x:2 dilation.z:1 dilation.y:1 dilation.x:1, 1, 3.3, 8) (Failed)
        10794 - Full/GPU_GroupConv3D_BackwardWeights_FP16.GroupConv3D_BackwardWeights_half_Test/( G:1 N:6 C:448 K:896 D:3 H:118 W:182 z:1 y:1 x:1 pad.z:0 pad.y:0 pad.x:0 stride.z:1 stride.y:2 stride.x:2 dilation.z:1 dilation.y:1 dilation.x:1, 1, 3.3, 7) (Failed)
        10795 - Full/GPU_GroupConv3D_BackwardWeights_FP16.GroupConv3D_BackwardWeights_half_Test/( G:1 N:6 C:448 K:896 D:3 H:118 W:182 z:1 y:1 x:1 pad.z:0 pad.y:0 pad.x:0 stride.z:1 stride.y:2 stride.x:2 dilation.z:1 dilation.y:1 dilation.x:1, 2.2, 0, 8) (Failed)
        10796 - Full/GPU_GroupConv3D_BackwardWeights_FP16.GroupConv3D_BackwardWeights_half_Test/( G:1 N:6 C:448 K:896 D:3 H:118 W:182 z:1 y:1 x:1 pad.z:0 pad.y:0 pad.x:0 stride.z:1 stride.y:2 stride.x:2 dilation.z:1 dilation.y:1 dilation.x:1, 2.2, 0, 7) (Failed)
        10797 - Full/GPU_GroupConv3D_BackwardWeights_FP16.GroupConv3D_BackwardWeights_half_Test/( G:1 N:6 C:448 K:896 D:3 H:118 W:182 z:1 y:1 x:1 pad.z:0 pad.y:0 pad.x:0 stride.z:1 stride.y:2 stride.x:2 dilation.z:1 dilation.y:1 dilation.x:1, 2.2, 3.3, 8) (Failed)
        10798 - Full/GPU_GroupConv3D_BackwardWeights_FP16.GroupConv3D_BackwardWeights_half_Test/( G:1 N:6 C:448 K:896 D:3 H:118 W:182 z:1 y:1 x:1 pad.z:0 pad.y:0 pad.x:0 stride.z:1 stride.y:2 stride.x:2 dilation.z:1 dilation.y:1 dilation.x:1, 2.2, 3.3, 7) (Failed)

For example, these are all ~96 GBs in size with the new calculation (1 x 448 x 896 x 3 x 118 x 182 x 4).

Note that this is fixed with latest changes.

JonathanLichtnerAMD committed:
Fix a couple of problems with the CK and subbuffer OOB checking integration

The first problem was that a zero size was being passed in and so the
subbuffer OOB checking code was allocating memory of size zero.  (This
workd fine if we're just using a subbuffer directly, since that will
reference inside of the workspace.)

The second problem was that the buf_handle moved out of scope after
the if statement, which triggered the allocated sub-buffer to be freed!

JonathanLichtnerAMD committed:
Use the weights tensor to calculate the CK workspace size

JonathanLichtnerAMD committed:
Reformat code

JonathanLichtnerAMD committed:
Add unit test to reproduce Fremont issue

JonathanLichtnerAMD committed:
Remove unneeded arguments from miopenConvolutionABBackwardWeightsGetWorkSpaceSize

BrianHarrisonAMD commented on src/include/miopen/solver/implicitgemm_ck_util.hpp:
```diff
@@ -1167,16 +1168,19 @@ ConvSolution InitInvokerFactoryNCHW(const ExecutionContext& ctx,
                                  CastType>(ck_args, sh_conv_ptr, tr_ptrs, data_ctx, split_k);
 
+            shared<Data_t> buf_handle{};
             if(ck_buff_des.has_value() && ck_buff_des->ck_size && workspace_ptr)
             {
-                auto buf_handle = handle.CreateSubBuffer(workspace_ptr, ck_buff_des->ck_offset, 0);
+                buf_handle = handle.CreateSubBuffer(
+                    workspace_ptr, ck_buff_des->ck_offset, ck_buff_des->ck_size);
```
Any reason the buf_handle had to move?

JonathanLichtnerAMD replied:
Yes, because buf_handle would go out of scope, which caused the allocated subbuffer to be freed (in the case where we were actually allocating a subbuffer for OOBs checking)

BrianHarrisonAMD replied:
It'll depend if its being ran with profiling / timing or not, but since this runs Async without timing enabled it could still be freed before the execution finishes.

BrianHarrisonAMD replied:
We may need to force flushing before freeing the specially allocated buffers if it's enabled to ensure it doesn't have this issue.

JonathanLichtnerAMD replied:
I would like to get this PR in for Fremont, so I can create an issue for this problem.  This only happens when we are running with the OOBs checking on sub-buffers, and can be worked around by using -t 1 (which is how I always run the MIOpenDriver commands, so this might be I never noticed this potential race condition before).

BrianHarrisonAMD commented on src/include/miopen/solver/implicitgemm_ck_util.hpp:
```diff
@@ -1167,16 +1168,19 @@ ConvSolution InitInvokerFactoryNCHW(const ExecutionContext& ctx,
             auto invoker_ptr = sh_conv_ptr->MakeInvokerPointer();
             {
                 WorkAroundHipEventProfiler prf(handle);
+                MIOPEN_LOG_I2("kernel_name = " << kernel_id);
```
Nice!

BrianHarrisonAMD commented on src/convolution_api.cpp:
```diff
@@ -266,15 +264,22 @@ miopenConvolutionABBackwardWeightsGetWorkSpaceSize(const miopenAlphaBetaCase_t a
             case miopenInt64: byte_size = 8; break;
             }
-            *buffer_size = byte_size * output_tensor_size;
+            *buffer_size = G * weights_tensor_size * byte_size;
```
GetNumBytes() is doing basically the byte * tensor_size calc, and could be used.
The fact that group count isn't included in GetElementSize, or GetNumBytes, is potentially a bit concerning though.
We should investigate where these are used, and ensure they don't need to account for group as well.

JonathanLichtnerAMD replied:
I took a look at GetNumBytes and it mostly seemed to be used when we needed to allocate space for the given tensor, e.g., something like

        const auto space = tensor_add_op.tensor_desc.GetNumBytes();
        auto ptr         = allocate_buffer(space);

I didn't see any uses of GetNumBytes with the workspace size calculations.  I also checked the usage of GetElementSize, and this seemed benign.

Anyway,  I can change this to use GetNumBytes instead of GetElementSize.

BrianHarrisonAMD replied:
Well I think it might be the case that group count is part of the tensor dims, and we don't need to add G * here, but that also means we would have to recreate the dims / strides to match what CK is expecting, and I see some logic for that in the CK solvers.

I think we need to review how grouped convs impact dims / strides inside MIOpen, and then track that all the way through to make sure it's consistently handled correctly.

JonathanLichtnerAMD replied:
Okay, I figured this out now.  Initially, I thought the weight tensor was {K, C, Z, Y, X} which is why I multiplied by G.  However, MIOpen divides K and C by G for CK to get K' and C'.  However the weight tensor is actually {GxK', C', Z, Y, X}, and so there is no need to multiply by G again since G is already in there.

This has passed CI and passed for the Fremont shape.  I've also added some additional unit tests for G > 1.

I will try GetNumBytes vs. GetElementSize and will make the switch if it passes testing.

JonathanLichtnerAMD replied:
PS I tried switching to GetNumBytes() from GetElementSize() but this caused memory access faults in the Fremont shape and many unit tests.  The fremont case was bfp16, but the old code was still multiplying by 4, so I suspect CK must be using floats internally in this workspace, i.e., we used to multiply weights_tensor_size * 4 to get the buffer_size

BrianHarrisonAMD commented on test/gtest/group_conv.hpp:
```diff
@@ -253,6 +253,7 @@ struct GroupConvTestConfig<3u>
                 {4, 128,  4 , 8  ,{28, 28, 28} , {3, 3, 3}, {1, 1, 1}, {2, 2, 2}, {1, 1, 1}},
                 {2, 128,  2 , 2  ,{28, 28, 28} , {3, 3, 3}, {1, 1, 1}, {2, 2, 2}, {1, 1, 1}},
                 {1, 6  , 448, 896,{3, 118, 182}, {1, 1, 1}, {0, 0, 0}, {1, 2, 2}, {1, 1, 1}},
+                {1, 256,  3 ,1280,{2,  14, 14} , {2,14,14}, {0, 0, 0}, {2,14,14}, {1, 1, 1}},
```
We probably should add a case with g > 1 as well, and may want to add some extra coverage for FWD as well.

JonathanLichtnerAMD replied:
I've added some additional tests

BrianHarrisonAMD commented on src/convolution_api.cpp:
```diff
@@ -218,23 +218,21 @@ extern "C" miopenStatus_t miopenGetConvolutionFindMode(const miopenConvolutionDe
-                                                   const miopenTensorDescriptor_t outputTensorDesc,
+                                                   const miopenTensorDescriptor_t weightsTensorDesc,
                                                    const miopenConvolutionDescriptor_t convDesc,
                                                    size_t* buffer_size)
```
Is this directly callable through the public API? 

I don't think this is properly accounting for the fact that it might need additional workspace if the input tensors are NCHW layout and it needs to have room to re-arrange the tensor prior to calling CK.

JonathanLichtnerAMD replied:
This is an internal api only.  miopenConvolutionABBackwardWeightsGetWorkSpaceSize() is defined in miopen_internal.h and only gets called by GetCKAlphaBetaWorkspace(), which is only called by various solvers and InitInvokerFactoryNCHW.  I think the other tensors have their size accounted for separately

BradPepersAMD reviewed (commented):
Brian covered all my concerns so I'm good to merge once those comments are addressed and we have passing CI.  need to also do whatever manual testing to increase confidence too.

JonathanLichtnerAMD committed:
Merge branch 'develop' into jlichtne/fix-workspace-calculation-for-CK

JonathanLichtnerAMD committed:
Add some additional unit tests for fwd and g > 1

JonathanLichtnerAMD committed:
Revert "Remove unneeded arguments from miopenConvolutionABBackwardWeightsGetWorkSpaceSize"

This reverts commit 9bce2c8db4927536227cf4c3b144de87206d52ad.

JonathanLichtnerAMD commented:
> Brian covered all my concerns so I'm good to merge once those comments are addressed and we have passing CI. need to also do whatever manual testing to increase confidence too.

@BradPepersAMD  Yes, I will test with a bunch of KTN shapes and I will also run CI for MI100 and MI300.

It might also be good to run this on 3D shapes:  do you have a good list of 3D shapes?

JonathanLichtnerAMD commented:
I ran this with 1500 backwards weighted shapes (from the KTN list that I had, randomly assorted) and this passed as before.

JonathanLichtnerAMD committed:
Do not multiply the weight tensor by G

For example, with G=3, C1=9, K1=27, Z=2, Y=14, and X=14, then MIOpen
computes K to be K1/G and C to be C1/G.  However, the weight tensor
becomes {27, 3, 2, 14, 14} instead of {9, 3, 2, 14, 14}, since K is
multiplied by G in the first element.

JonathanLichtnerAMD committed:
Add some more unit tests for g > 1

BrianHarrisonAMD reviewed (approved):
LGTM, can probably do that one swap to using elementspace, but otherwise looks correct to me now.

JonathanLichtnerAMD committed:
Add a comment indicating why GetNumBytes() cannot be used

JonathanLichtnerAMD committed:
Merge branch 'develop' into jlichtne/fix-workspace-calculation-for-CK
